### PR TITLE
[flang][OpenMP] Fix for #86393

### PR DESCRIPTION
**Pull request by SouraVX**
When checking symbol details:
```
c (OmpReduction): HostAssoc

### CODE REVIEW

**Review by tblah — COMMENTED** [review]
Thank you for fixing this. The changes to generated IR look good to me. Please could you add a test similar to `flang/test/Lower/OpenMP/parallel-reduction-array.f90` which covers this case so that it does not get missed in the future.

(This shouldn't block the patch, but) I don't fully understand why skipping one level in the symbol table results in us using the correct result from the hlfir.declare. Please could you explain this a bit more for me?

**Comment by SouraVX** [comment]
> Thank you for fixing this. The changes to generated IR look good to me. Please could you add a test similar to `flang/test/Lower/OpenMP/parallel-reduction-array.f90` which covers this case so that it does not get missed in the future.
Okay
> (This shouldn't block the patch, but) I don't fully understand why skipping one level in the symbol table results in us using the correct result from the hlfir.declare. Please could you explain this a bit more for me?

Something has to do with:
https://github.com/llvm/llvm-project/blob/main/flang/lib/Lower/Bridge.cpp#L1031

File: flang/test/Lower/OpenMP/parallel-reduction3.f90
Comment (tblah) [inline flang/test/Lower/OpenMP/parallel-reduction3.f90:1]: It is better not to hardcode the names of SSA values (e.g. `"%0"`).

You can automatically generate most of these CHECK lines with https://github.com/llvm/llvm-project/blob/main/mlir/utils/generate-test-checks.py

File: flang/lib/Lower/OpenMP/ReductionProcessor.cpp [hunk 1]
```diff
@@ -527,7 +528,10 @@ void ReductionProcessor::addDeclareReduction(
     // all arrays must be boxed so that we have convenient access to all the
     // information needed to iterate over the array
     if (mlir::isa<fir::SequenceType>(redType.getEleTy())) {
-      hlfir::Entity entity{symVal};
+      // For Host associated symbols, use `SymbolBox` instead
+      Fortran::lower::SymbolBox symBox =
+          converter.lookupOneLevelUpSymbol(*symbol);
```
Comment (tblah) [inline flang/lib/Lower/OpenMP/ReductionProcessor.cpp:533]: Why does looking one level up always work, even in cases where the symbol is not host associated (e.g. in the tests modified by this PR)?

**Review by tblah — APPROVED** [review]
Okay so my understanding is that lookOneLevelUp will always get the host symbol from the function scope. Previously we were using the privatized version of the symbol from the OpenMP construct scope.

It isn't clear to me why the OpenMP construct scope maps the symbol to the second result of the hlfir.declare, meanwhile the function scope uses the first result of the declare (which then gives the right result here). I wonder if that is an OpenMP bug.

But as this is a crash seen in the wild, I do not want to hold up this PR further.

Thank you for fixing this.

**Comment by kiranchandramohan** [comment]
> It isn't clear to me why the OpenMP construct scope maps the symbol to the second result of the hlfir.declare, meanwhile the function scope uses the first result of the declare (which then gives the right result here). I wonder if that is an OpenMP bug.

I think `getSymbolAddress` returns the `#1` value. Are you suggesting that getting the `#0` value is the right fix here and not using `lookoneLevelUp` ?

**Comment by tblah** [comment]
> I think `getSymbolAddress` returns the `#1` value. Are you suggesting that getting the `#0` value is the right fix here and not using `lookoneLevelUp` ?

As I understand it, this is what lookOneLevelUp is resulting in, which is what fixes the bug here (the `#0` value is a box whereas `#1` is a reference).

**Pull request merged**